lib/configuration.js

```diff
@@ -15,150 +15,615 @@
+  /**
+   * The _givenConfiguration property holds a ConfigurationOptions object
+   * which, if valid, will be merged against by the values taken from the meta-
+   * data service. If the _givenConfiguration property is not valid then only
+   * metadata values will be used in the Configuration instance.
+   * @memberof Configuration
+   * @private
+   * @type {Object|Null}
+   * @defaultvalue null
+   */
+  this._givenConfiguration = isPlainObject(givenConfig) ? givenConfig : null;
+  /**
+   * The _logger property holds a logger instance which, if valid, will be used
+   * to log any debug or error information generated by the instance.
+   * @memberof Configuration
+   * @private
+   * @type {Object|Null}
+   * @defaultvalue null
+   */
+  this._logger = isObject(logger) ? logger : null;
+}
+// Extend the Configuration constructor by augmenting it with EventEmitter
+inherits(Configuration, EventEmitter);
 /**
- * Attempts to determine whether or not a service context parameter was given
- * by the Object configuration. If it is and it is of the correct type then
- * it will be merged with the default service context where the given
- * configuration will overwrite all shared properties on the default return
- * value.
- * @function extractServiceContext
- * @param {ConfigurationOptions} initConfiguration - the init configuration
- * @returns {ServiceContextConfiguration} - The service context of the hosting
- *  application
- */
-function extractServiceContext(initConfiguration) {
-
-  var gaeModuleName = env.GAE_MODULE_NAME;
-  var gaeModuleVersion = env.GAE_MODULE_VERSION;
-
-  var defaultReturnObject = {
-    service : isString(gaeModuleName) ? gaeModuleName : '',
-    version : isString(gaeModuleVersion) ? gaeModuleVersion : ''
-  };
-
-  if (has(initConfiguration, 'serviceContext') &&
-      isPlainObject(initConfiguration.serviceContext)) {
-
-    if (has(initConfiguration.serviceContext, 'service')) {
-
-      defaultReturnObject.service = initConfiguration.serviceContext.service;
-    }
-
-    if (has(initConfiguration.serviceContext, 'version')) {
-
-      defaultReturnObject.version = initConfiguration.serviceContext.version;
+ * The callback for `utils.getProjectNumber` this function is responsible for
+ * determining whether or not an error occured in the metadata transaction and,
+ * if one did not, assigning the value of the transaction (the project number)
+ * to the instance property `_projectNumber`.
+ * @memberof Configuration
+ * @private
+ * @function _assimilateProjectNumber
+ * @param {Error|Undefined} err - If applicable, the error that occured during 
+ *  the transaction
+ * @param {Null|String} projectNum - The project number if successfully 
+ *  retrieved, null otherwise.
+ * @returns {Undefined} - does not return anything
+ */
+Configuration.prototype._assimilateProjectNumber = function(err, projectNum) {
+  if (isNull(err) && isString(projectNum)) {
+    this._projectNumber = projectNum;
```

Comment:
@matthewloring That's a good idea, should I land that as a `console.log` in this commit or should I wait till my next PR - I was planning on the next PR swapping out all the logging stuff to the common diagnostics logging stuff